### DISCUSSION TITLE
Logging Design and Improvements

### DISCUSSION THREAD

**Discussion by andfranklin** [terrapower/armi#1812] [General]
This is a spin-off of the conversation started in issue #1807. A brief overview of my view:

- There isn't great guidance on all of the [additional logging levels](https://github.com/terrapower/armi/blob/e2ab9418f3c3b3b95d2f386ad671611452760622/armi/runLog.py#L120-L131) that are defined in addition to the [logging levels provided with python by default](https://docs.python.org/3/library/logging.html#logging-levels). This can lead to confusion for analysts and [developers](https://github.com/terrapower/armi/issues/1807#issuecomment-2266629449). [Some guidance](https://github.com/terrapower/armi/pull/1810) is in the works, but it seems targeted at developers. Having more complicated logging levels will require more complicated guidance -- which I would like to avoid.
- [Some of the design decisions](https://github.com/terrapower/armi/blob/e2ab9418f3c3b3b95d2f386ad671611452760622/armi/runLog.py#L689-L691) make it difficult to work with ARMI applications, and they don't seem necessary.

The big concern seems to be the amount of effort required to implement the changes, and the unknown consequences of making changes. Admittedly I don't have a complete perspective on all of ARMI's use cases, and I hope that starting a discussion will unsurface other people's perspectives. My hope is that we can trim-down the ARMI's `runLog` requirements to the bare essential, the design can be simplified, and guidance can be provided to developers and analysts.

**Comment by albeanth** [comment]
> There isn't great guidance on all of the additional logging levels that are defined...

I'd generally agree to this and can see how this would be interpreted as a documentation gap. At a quick glance, it looks like this is all there is in the dev manual for logs. 
https://terrapower.github.io/armi/developer/tooling.html#module-level-logging
What's provided in #1810 is a great addition, no doubt. Maybe it is sufficient to fill the gaps, maybe not. I'd be curious to know what the other armi devs think.

**Comment by jakehader** [comment]
> * [Some of the design decisions](https://github.com/terrapower/armi/blob/e2ab9418f3c3b3b95d2f386ad671611452760622/armi/runLog.py#L689-L691) make it difficult to work with ARMI applications, and they don't seem necessary.

Hi @andfranklin - Do you have a simple example of how this run log choice makes it difficult to work with? This is the best video / documentation I've found on pythons logging - https://youtu.be/9L77QExPmI0?si=TbFYHYvZX6MtHkvB

**Comment by john-science** [comment]
> * There isn't great guidance on all of the additional logging levels.

I have a PR open for this: https://github.com/terrapower/armi/pull/1810

(I can't MERGE that PR right now, because we are in a merge freeze. But soon.)

> * [Some of the design decisions](https://github.com/terrapower/armi/blob/e2ab9418f3c3b3b95d2f386ad671611452760622/armi/runLog.py#L689-L691) make it difficult to work with ARMI applications, and they don't seem necessary.

As Jake said, what is your specific feature request?
